src/stream/src/executor/merge.rs

```diff
@@ -464,6 +477,92 @@ impl SelectReceivers {
     }
 }
 
+/// A wrapper that buffers the `StreamChunk`s from upstream until no more ready items are available.
+/// Besides, any message other than `StreamChunk` will trigger the buffered `StreamChunk`s
+/// to be emitted immediately, as well as the message itself.
+struct BufferChunks<S: Stream> {
+    inner: S,
+    chunk_builder: StreamChunkBuilder,
+
+    /// The items to be emitted. Whenever there's something here, we should return a `Poll::Ready` immediately.
+    pending_items: VecDeque<S::Item>,
+}
+
+impl<S: Stream> BufferChunks<S> {
+    pub(super) fn new(inner: S, chunk_size: usize, schema: Schema) -> Self {
+        assert!(chunk_size > 0);
+        let chunk_builder = StreamChunkBuilder::new(chunk_size, schema.data_types());
+        Self {
+            inner,
+            chunk_builder,
+            pending_items: VecDeque::new(),
+        }
+    }
+}
+
+impl<S: Stream> std::ops::Deref for BufferChunks<S> {
+    type Target = S;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<S: Stream> std::ops::DerefMut for BufferChunks<S> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.inner
+    }
+}
+
+impl<S: Stream> Stream for BufferChunks<S>
```

Comment:
I think we can use the `try_stream` macro and `yield` to avoid implementing a stateful stream on our own. We can use `self.inner.try_next().now_or_never()` to get the ready chunk without blocked at await and also be aware of the pending.

Reply:
I wish I had known this before working out the PR 😌

Reply:
This might be helpful:

https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.ready_chunks

Reply:
Yeah, actually I just followed the `ready_chunks` to write `BufferChunks`. `ready_chunks` can work here, but my self-implemented code can do better - it buffers chunks purely according to our customized logic i.e. when `ChunkBuilder` is full or when barrier/watermark comes. Meanwhile for `ready_chunks`, we might need to set a large enough buffer size and do the chunking outside.